clippy_lints/src/matches/significant_drop_in_scrutinee.rs

```diff
@@ -22,6 +22,13 @@ pub(super) fn check<'tcx>(
         for found in suggestions {
             span_lint_and_then(cx, SIGNIFICANT_DROP_IN_SCRUTINEE, found.found_span, message, |diag| {
                 set_diagnostic(diag, cx, expr, found);
+                diag.span_label(
+                    expr.span,
+                    "temporaries in a match scrutinee are not dropped until the end of the \
+                    match, so side effects from dropping such temporaries are delayed. \
+                    Deadlocks can occur if a lock guard is created in the scrutinee and a new \
+                    lock guard for the same mutex is created in a match arm. \
+                    For more information, see https://rust-lang.github.io/rust-clippy/master/index.html#significant_drop_in_scrutinee");
```

Comment:
This link is automatically added to every clippy lint, unless explicitly disabled through an undocumented env var. So no need to add it here.

I don't think adding such a big paragraph to a lint message is what was suggested. We don't add full lint explanations to lint messages, that's what the lint documentation is for. I can't think of a rustc or Clippy lint where something like this is done. Can you link the discussion where adding this was suggested?

Reply:
@flip1995 Thank you for the review! It looks like I might have misunderstood Oli's suggestion here: https://github.com/rust-lang/rust/issues/93883#issuecomment-1122007881

Particularly this part:
```
the diagnostic messaging needs to explain why this is a problem, ideally with some links to documentation and examples

    maybe even add a short span pointing to the end of the match stating that the lock holds until there, so if any arms directly or via function calls lock again, you're in a deadlock
```

I've tried to make the messages smaller, but I did notice some PRs opened due to this lint where it was mentioned that the current output was unhelpful and confusing. If you think this is still too much, I will keep trying to trim it down.

Reply:
> This link is automatically added to every clippy lint, unless explicitly disabled through an undocumented env var. So no need to add it here.

Oh, I completely missed this! Thank you! The URL has been removed as well.

Reply:
Ah yeah, we might want to improve the messaging a bit. But just writing a big paragraph won't help here. I think the ideal thing to do would be something like:

```console
error: temporary with drop impl with side effects in match scrutinee lives to end of block
  --> $DIR/significant_drop_in_scrutinee.rs:59:11
   |
LL |     match mutex.lock().unwrap().foo() {
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
LL |         42 => mutex.lock().unwrap().bar()
    |              ^^^^^^^^^^^^ note: new side effect temporary created here
...
LL |     };
   |      - original temporary lives until here
   |
   = note: this might lead to deadlocks or other unexpected behavior
help: try moving the temporary above the match
   |
LL ~     let value = mutex.lock().unwrap().foo();
LL ~     match value {
   |
```

But this would entail, that we also try to find other significant drop temporaries inside the match block, which might be a bit hard.